### PR TITLE
Use CLFLUSH when CLFLUSHOPT is not available

### PR DESCRIPTION
**Pull request by OmSaran**
Use the compiler argument -march=native to know if CLFLUSHOPT is supported.

The function name might be misleading, so if need be I can go ahead and replace all the calls with a macro.

### CODE REVIEW

**Comment by vijay03** [comment]
@OmSaran: thanks for the pull request! I've heard that sometimes march=native results in code breaking as it enables a bunch of flags. Is there a way to find out if CLFLUSHOPT is supported in a more fine-grained manner?

File: splitfs/non_temporal.h [hunk 1]
```diff
@@ -50,8 +50,13 @@
 #define MOVNT_THRESHOLD_GRANULARITY 64
 
 #define CLFLUSH_SIZE 64
-#define _mm_clflushopt(addr)\
-	asm volatile("clflushopt %0" : "+m" (*(volatile char *)(addr)))
+#ifndef __CLFLUSHOPT__
+ #define _mm_clflushopt(addr)\
```
Comment (vijay03) [inline splitfs/non_temporal.h]: We probably want to rename this to something other than _mm_clflushopt -- maybe _mm_flush?

File: README.md [hunk 2]
```diff
@@ -89,7 +89,7 @@ The kernel patch for the implementation of relink() system call for linux v4.13
 2. At least 32 GB DRAM
 3. At least 4 cores
 4. Baremetal machine (Not a VM)
-5. Intel Processor supporting `clflushopt` instruction (Introduced in Intel processor family -- Broadwell). This can be verified with `lscpu | grep clflushopt`
+5. Intel Processor supporting `clflush` or `clflushopt` (Introduced in Intel processor family -- Broadwell) instruction. This can be verified with `lscpu | grep clflush`. `clflushopt` will lead to better performance in strict mode.
```
Comment (vijay03) [inline README.md]: I think we also want to add a new section "Contributors" near the end of README and add your name and github handle there.

**Comment by rohankadekodi** [comment]
> @OmSaran: thanks for the pull request! I've heard that sometimes march=native results in code breaking as it enables a bunch of flags. Is there a way to find out if CLFLUSHOPT is supported in a more fine-grained manner?

Can we check for clflushopt / clwb support using something like this?
[clflushopt_support.txt](https://github.com/utsaslab/SplitFS/files/4342305/clflushopt_support.txt)

**Comment by OmSaran** [comment]
@vijay03 @rohankadekodi
Thanks for the review!

I tried looking for a more granular check on the compiling system. But I was unable to find.

However, I can think of 2 ways of approaching this without having a conditional check every time we want to execute the flush instruction:
1. Determine at compile time - Write a new file specifically for the flush function and isolate its compilation with march=native
2. Determine at run time - During the start up of SplitFS determine the appropriate instruction using a function pointer - would this be the right place to add it? - https://github.com/utsaslab/SplitFS/blob/master/splitfs/fileops_nvp.c#L1101 

For 1) I'm assuming that the way users will use SplitFS is by compiling it themselves and not rely on pre-built artefacts. 

Also, I was thinking of leveraging builtin functions https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/cpuid.h#L196 to implement the check (in case of runtime check).

What do you guys suggest?

**Comment by rohankadekodi** [comment]
Hi @OmSaran, @vijay03 

Could you check the code snippet that I have attached here to check whether CLFLUSHOPT is supported? 

We can have a global variable (bool clflushopt_support) that you will set in the _nvp_init2()  function (https://github.com/utsaslab/SplitFS/blob/master/splitfs/fileops_nvp.c#L1101), by checking in a way similar to how this code snippet checks it. Then based on the value of the global variable, we can either use clflushopt or clflush in the code. 

[clflushopt_support.txt](https://github.com/utsaslab/SplitFS/files/4345557/clflushopt_support.txt)

**Comment by vijay03** [comment]
So the code snippet from @rohankadekodi is this:

```
#include <stdio.h>
#include <stdlib.h>
#include <immintrin.h>
#include <emmintrin.h>

#define bit_CLFLUSHOPT (1 << 23)
#define bit_CLWB (1 << 24)

#define __cpuid_count(level, count, a, b, c, d)     \
    __asm__ ("cpuid\n\t"                  \
            : "=a" (a), "=b" (b), "=c" (c), "=d" (d) \
            : "0" (level), "2" (count))

void main() {
    unsigned eax, ebx, ecx, edx;
    __cpuid_count (7, 0, eax, ebx, ecx, edx);
    char buffer[64];
    if(ebx & bit_CLFLUSHOPT)
    printf("supprts clflushopt\n");
}
```
Assuming this works, I think using this over march=native to find if there is CLFUSHOPT support is a good idea.

**Comment by vijay03** [comment]
Thanks for making the changes, @OmSaran!

File: splitfs/non_temporal.h (hunk 1 above)
Comment (OmSaran) [reply splitfs/non_temporal.h]: Done

File: README.md (hunk 2 above)
Comment (OmSaran) [reply README.md]: Thank you! Done

**Pull request merged**